Time based cover with tilt function

klaudiusz223 opened the pull request:
# What does this implement/fix?

Time based cover implementation with position and tilt control that are driven by one motor and do not have any position feedback. Such a solution can be found in the external venetian blinds.
Has an automatic recalibration function at extreme positions (fully open or fully closed). Takes into account the inertia time and the interlocking time when changing direction.
It can also be used to control covers without a tilt function. In such case, the rotation time must be set to 0. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

https://github.com/esphome/feature-requests/issues/664
https://github.com/esphome/feature-requests/issues/2537
https://github.com/esphome/feature-requests/issues/2087
https://github.com/esphome/feature-requests/issues/1329
https://github.com/esphome/feature-requests/issues/778
https://github.com/esphome/feature-requests/issues/739

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 
   esphome/esphome-docs#3574
Preview: https://deploy-preview-3574--esphome.netlify.app/components/cover/time_based_tilt/

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration entry
cover:
  - platform: time_based_tilt
    name: "Time-Based Tilt Cover"
    open_duration: 58700ms
    close_duration: 57700ms
    inertia_open_time: 300ms
    tilt_open_duration: 930ms
    inertia_close_time: 250ms
    tilt_close_duration: 900ms
    recalibration_close_time: 6500ms
    recalibration_open_time: 4500ms
    actuator_activation_open_time: 200ms
    actuator_activation_close_time: 250ms
    close_sets_tilt: false
    open_sets_tilt: false

    open_action:
      - switch.turn_on: open_cover_switch

    close_action:
      - switch.turn_on: close_cover_switch

    stop_action:
      - switch.turn_off: open_cover_switch
      - switch.turn_off: close_cover_switch

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

klaudiusz223 committed:
implementation of time based cover with tilt

klaudiusz223 committed:
- fix publishing state
- use constants for open and close posititions

klaudiusz223 committed:
fixed recalibration if covers w/o tilt function

klaudiusz223 committed:
fixes compilation error with esp-idf

klaudiusz223 committed:
cleaning

klaudiusz223 committed:
cleaning

klaudiusz223 committed:
Fixing a bug when issuing the stop command just after the start (between the transition from STATE_IDLE to STATE_MOVING).
Now the STOP command always stops the cover even during recalibration.

klaudiusz223 committed:
set supports stop

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

klaudiusz223 committed:
added some comments

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

klaudiusz223 committed:
static const

klaudiusz223 committed:
add test

klaudiusz223 committed:
update CODEOWNERS

klaudiusz223 committed:
fix formatting

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

klaudiusz223 committed:
cleaning up

MichalLytek commented:
I'm also eager to see this in action 😍 
My Shelly 2 PM are waiting for this 😄 

@klaudiusz223 Would it be hard to use the power meter sensors like in the `current_based` cover platform to distinguish between tilt and move/cover? 🤔

klaudiusz223 commented:
Hi @DaniEll-AT 
Thanks for the feedback.
please describe in more details what does it mean doesn't work and how do You test small steps and what is the small step.
Doesn't move at all? Or move to the wrong position?
Do You control it localy eg by button or do You send command by Home Assistant. And what command? Do You take into accout inertia?
And Yes, there is a rounding to 1%, for both  tilt and postion.
But most important is what is Your loop time in the esphome.
e.g  (for 900 ms tilt time )   20 ms means that the smallest step will be greater than 2% 
 30 ms means more than 3%
but if You want to tilt by 5% it will be  probably reported as tilt by not less than 7% (3x 2.22% - and rounding).
I thought also about implementing tilting using HighFrequencyLoopRequester

MichalLytek commented:
> What is exactly Your case scenario. Why do You need to distinguish tilting from moving? is the power consuption different for moving and tilting?

Yes, there's even a different noise of the engine while tilting and moving. I'm afraid of the tricky time measuring inertia and tilting time, would be nice to use current reading to have auto calibration done this way.

But first let me try your code on my Shelly's 😉

Bascht74 commented:
> I will consider whether it would be easy to implement what You need because I wanted to simplify my implementation a bit

Hi, @klaudiusz223,

anything new about simplifying or adding that feature? 
I would love to see that features on my Shelly devices so I am hoping you can finish the work soon so that it can be merged (milestone was changed)

maromme commented:
@klaudiusz223 thank's for your great work. it works for me flawlessly. Didn't find any problemes. 
Maybe one improvement: It would be nice if there was a option to choose weather the cover_recalibration_time is added at the top or bottom endstop. I've got a blind that should stop 1s bevore the bottom endstop and still I'd like the recalibration time at the top endstop.

klaudiusz223 committed:
Revert "add test"

This reverts commit a6b25f1eefb98c45b9ac0a1bd819e631059ad52d.

klaudiusz223 committed:
Merge branch 'dev' into tilt-cover

klaudiusz223 committed:
Revert "add test"

This reverts commit a6b25f1eefb98c45b9ac0a1bd819e631059ad52d.

klaudiusz223 committed:
Merge branch 'esphome:dev' into pcf_tilt

klaudiusz223 committed:
Refactoring – simplified.
Added actuator activation time.
Separated recalibration times for opening and closing.
Rounded position after stopping; saved rounded position and tilt values.
Added debug logging.
Stopped as needed in calibration state after receiving control requests.

klaudiusz223 committed:
Merge branch 'pcf_tilt' into tilt-cover

klaudiusz223 commented:
I've made some improvements to the code.
I added actuator activation time – separately for closing and opening. @DaniEll-AT @Bascht74 
I split the calibration process for the top and bottom positions. @maromme Is that what you meant?

Bascht74 commented:
Not quite. 

The calibration I want (To set manually) is that motors get power but doesn’t start right away or slowing down at the end positions. 
So the movement time of the cover doesn’t reflect the exact position. So what is needed is a linear correction of that. You should be able to say that e.g. 95% of the opening time is fully open 95–>100), and after 50% of the opening time it is allready at a 65% position (50->65) and at 10% of the opening time is is at a 5% position (10->5) and after 1% of the opening time it really starts opening (1->0). It would be the same that tasmota implemented here:

https://youtu.be/Z-grrvnu2bU?si=73ERqEmoKLBMrtQ5

https://tasmota.github.io/docs/Blinds-and-Shutters
(Look after Increasing Calibration Granularity)

That would help a lot of people with strange shutters (like the German ones behind the windows) and in other situations wheee the running time doesn’t mechanically reflect the position of the cover.

Sebastian

klaudiusz223 committed:
The cover, when in an extreme position, will enter calibration mode if it is fully open and receives the "open" command, or if it is fully closed and receives the "close" command.

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

klaudiusz223 committed:
add tests for time_based_tilt

sleepymaxx reviewed (approved):
Thank´s for your Work really apreciate it!

sleepymaxx commented on CODEOWNERS:
```diff
@@ -423,6 +423,7 @@ esphome/components/template/fan/* @ssieb
 esphome/components/text/* @mauritskorse
 esphome/components/thermostat/* @kbx81
 esphome/components/time/* @OttoWinter
+esphome/components/time_based_tilt/* @klaudiusz223
```
great work

XilviuX commented:
@klaudiusz223 not sure you're still maintaing your code but, as reported above, it gives a warning now:

```
WARNING Using `cover.COVER_SCHEMA` is deprecated and will be removed in ESPHome 2025.11.0. Please use `cover.cover_schema(...)` instead. If you are seeing this, report an issue to the external_component author and ask them to update it. 
```

Can you please update this?

klaudiusz223 committed:
Merge branch 'dev' into tilt-cover

klaudiusz223 committed:
[time_based_tilt] Update to use cover_schema(...)

klaudiusz223 committed:
update CODEOWNERS

klaudiusz223 committed:
Add close_sets_tilt and open_sets_tilt options to time_based_tilt cover

- Introduced new boolean config parameters:
  * close_sets_tilt — automatically set tilt to 0.0 when position=0.0 and no tilt is provided
  * open_sets_tilt — automatically set tilt to 1.0 when position=1.0 and no tilt is provided
- Behavior applies only to full open/close commands; partial positions remain unchanged

klaudiusz223 committed:
[time_based_tilt] Modernize to C++17 nested namespaces

klaudiusz223 committed:
Merge branch 'dev' into tilt-cover

klaudiusz223 committed:
[time_based_tilt] Corrected test component folder name to match the actual component ID.

klaudiusz223 committed:
[time_based_tilt] Fix tests

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover

bdraco commented on tests/components/time_based_tilt/test.esp32-ard.yaml:
```diff
@@ -0,0 +1 @@
+<<: !include common.yaml
```
Please remove duplicate tests for ard as there is no ard specific code in this PR

bdraco commented on tests/components/time_based_tilt/test.esp32-c3-ard.yaml:
```diff
@@ -0,0 +1 @@
+<<: !include common.yaml
```
Please remove duplicate tests for c3 as there is no c3 specific code in this PR

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.h:
```diff
@@ -0,0 +1,93 @@
+  bool is_at_target_tilt_() const;
+  bool is_at_extreme_position_() const;
+
+  Trigger<> *open_trigger_{new Trigger<>()};
+  Trigger<> *close_trigger_{new Trigger<>()};
+  Trigger<> *stop_trigger_{new Trigger<>()};
```
Please avoid heap allocations for Triggers

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.h:
```diff
@@ -0,0 +1,93 @@
+  Trigger<> *close_trigger_{new Trigger<>()};
+  Trigger<> *stop_trigger_{new Trigger<>()};
+
+  uint32_t open_duration_;
+  uint32_t close_duration_;
+
+  uint32_t tilt_close_duration_;
+  uint32_t tilt_open_duration_;
+
+  uint32_t interlock_wait_time_;
+  uint32_t recalibration_open_time_;
+  uint32_t recalibration_close_time_;
+  uint32_t inertia_open_time_;
+  uint32_t inertia_close_time_;
+  uint32_t actuator_activation_open_time_;
+  uint32_t actuator_activation_close_time_;
+
+  uint32_t current_recalibration_time_;
+  uint32_t current_actuator_activation_time_;
+
+  const static float TARGET_NONE;
+  enum State : uint8_t { STATE_IDLE, STATE_MOVING, STATE_STOPPING, STATE_CALIBRATING };
+
+  uint32_t last_recompute_time_{0};
+  uint32_t last_publish_time_{0};
+  float target_position_{TARGET_NONE};
+  float target_tilt_{TARGET_NONE};
+  float inertia_{0.0f};
+  bool has_built_in_endstop_{false};
+  bool assumed_state_{false};
+  cover::CoverOperation last_operation_{cover::COVER_OPERATION_OPENING};
+  State fsm_state_{STATE_IDLE};
+  cover::CoverOperation interlocked_direction_{cover::COVER_OPERATION_IDLE};
+  uint32_t interlocked_time_{0};
+  bool close_sets_tilt_{false};
+  bool open_sets_tilt_{false};
```
Please align members from largest to smallest

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+  if (this->fsm_state_ == STATE_IDLE && this->target_position_ == TARGET_NONE && this->target_tilt_ == TARGET_NONE)
+    return;
+
+  const uint32_t now = millis();
```
```suggestion
  const uint32_t now = App.get_loop_component_start_time();
```

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+  const uint32_t now = millis();
+
+  // recalibrating in extreme postions
+  if (this->fsm_state_ == STATE_CALIBRATING) {
```
Looks like this could be written as a switch on `this->fsm_state_`

bdraco replied:
For readability, I'd break each branch into its own function

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+
+void TimeBasedTiltCover::dump_config() {
+  LOG_COVER("", "Time Based Tilt Cover", this);
+  ESP_LOGCONFIG(TAG, "  Open Duration: %.3fs", this->open_duration_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Close Duration: %.3fs", this->close_duration_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Tilt Close Duration: %.3fs", this->tilt_close_duration_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Tilt Open Duration: %.3fs", this->tilt_open_duration_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Interlock wait time: %.3fs", this->interlock_wait_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Inertia close time: %.3fs", this->inertia_close_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Inertia open time: %.3fs", this->inertia_open_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Recalibration close time: %.3fs", this->recalibration_close_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Recalibration open time: %.3fs", this->recalibration_open_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Actuator activation close time: %.3fs", this->actuator_activation_close_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Actuator activation open time: %.3fs", this->actuator_activation_open_time_ / 1e3f);
+  ESP_LOGCONFIG(TAG, "  Close sets tilt: %s", this->close_sets_tilt_ ? "YES" : "NO");
+  ESP_LOGCONFIG(TAG, "  Open sets tilt: %s", this->open_sets_tilt_ ? "YES" : "NO");
+  ESP_LOGCONFIG(TAG, "  Current position: %.4f", this->position);
+  ESP_LOGCONFIG(TAG, "  Current tilt: %.4f", this->tilt);
```
Please combine ESP_LOGCONFIG statements

https://developers.esphome.io/architecture/logging/

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.h:
```diff
@@ -0,0 +1,93 @@
+  float target_position_{TARGET_NONE};
+  float target_tilt_{TARGET_NONE};
+  float inertia_{0.0f};
+  bool has_built_in_endstop_{false};
```
never used

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.h:
```diff
@@ -0,0 +1,93 @@
+  }
+  void set_close_sets_tilt(bool value) { this->close_sets_tilt_ = value; }
+  void set_open_sets_tilt(bool value) { this->open_sets_tilt_ = value; }
+  cover::CoverOperation compute_direction(float target, float current) {
```
should be const

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.h:
```diff
@@ -0,0 +1,93 @@
+    }
+    return target < current ? cover::COVER_OPERATION_CLOSING : cover::COVER_OPERATION_OPENING;
+  };
+  float round_position(float pos) { return round(100 * pos) / 100; };
```
should be const

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+
+const float TimeBasedTiltCover::TARGET_NONE = -1;
+
+using namespace esphome::cover;
```
```suggestion
```

Please qualify namespaces instead

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+  if (restore.has_value()) {
+    restore->apply(this);
+  } else {
+    this->position = 0.5f;
+    this->tilt = 0.5f;
```
please use a constexpr for magic numbers
INERTIA_CENTER = 0.5f

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+        this->publish_state(false);
+        ESP_LOGD(TAG, "Transition to the calibration state");
+      } else {
+        this->fsm_state_ = STATE_STOPPING;
+        ESP_LOGD(TAG, "Transition to the stopping state");
```
This pattern appears many times, please make a helper

```
  void TimeBasedTiltCover::transition_to_stopping_() {                                                                                                                                                                                                  
    this->fsm_state_ = STATE_STOPPING;                                                                                                                                                                                                                  
    ESP_LOGD(TAG, "Transition to the stopping state");                                                                                                                                                                                                  
  }            
```

bdraco commented on esphome/components/time_based_tilt/time_based_tilt_cover.cpp:
```diff
@@ -0,0 +1,356 @@
+  }
+}
+
+float TimeBasedTiltCover::get_setup_priority() const { return setup_priority::DATA; }
```
```suggestion
```

DATA is default

klaudiusz223 committed:
Merge branch 'esphome:dev' into tilt-cover
